cpp/src/c_api/replicate_edgelist.cpp

```diff
@@ -0,0 +1,157 @@
+
+    rmm::device_uvector<size_t> edge_offsets(2, handle.get_stream());
+    std::vector<size_t> h_edge_offsets{{0, gathered_edgelist_srcs.size()}};
+    raft::update_device(
+      edge_offsets.data(), h_edge_offsets.data(), h_edge_offsets.size(), handle.get_stream());
+
+    // FIXME: Hnadle this case better
+    if (edgelist_weights) {
+      auto gathered_edgelist_weights = cugraph::c_api::detail::device_allgatherv(
+        handle,
+        raft::device_span<weight_t const>(edgelist_weights->data(), edgelist_weights->size()));
+
+      *result = new cugraph::c_api::cugraph_induced_subgraph_result_t{
+        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_srcs, src->type_),
+        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_dsts, dst->type_),
+        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_weights,
+                                                               weights->type_),
+        new cugraph::c_api::cugraph_type_erased_device_array_t(edge_offsets,
+                                                               cugraph_data_type_id_t::SIZE_T)};
+    } else {
+      *result = new cugraph::c_api::cugraph_induced_subgraph_result_t{
+        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_srcs, src->type_),
+        new cugraph::c_api::cugraph_type_erased_device_array_t(gathered_edgelist_dsts, dst->type_),
+        NULL,
+        new cugraph::c_api::cugraph_type_erased_device_array_t(edge_offsets,
+                                                               cugraph_data_type_id_t::SIZE_T)};
+    }
+
+  } catch (std::exception const& ex) {
+    *error = new cugraph::c_api::cugraph_error_t{ex.what()};
+    return CUGRAPH_UNKNOWN_ERROR;
+  }
+
+  return CUGRAPH_SUCCESS;
+}
+
+}  // namespace
+
+// template <typename vertex_t, typename weight_t>
+//  why taking out the 'extern "C"' worked.
+extern "C" cugraph_error_code_t cugraph_replicate_edgelist(
+  const cugraph_resource_handle_t* handle,
+  const cugraph_type_erased_device_array_view_t* src,
+  const cugraph_type_erased_device_array_view_t* dst,
+  const cugraph_type_erased_device_array_view_t* weights,
+  cugraph_induced_subgraph_result_t** result,
+  cugraph_error_t** error)
+{
+  auto p_handle = reinterpret_cast<cugraph::c_api::cugraph_resource_handle_t const*>(handle);
+  auto p_src =
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(src);
+  auto p_dst =
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(dst);
+  auto p_weights =
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(weights);
+
+  /*
+  reinterpret_cast<cugraph::c_api::cugraph_induced_subgraph_result_t**>(result),
+  reinterpret_cast<cugraph::c_api::cugraph_error_t**>(error)
+  */
+
+  // FIXME: Support int64_t and float as well
```

Comment:
You should be able to call `cugraph::c_api::vertex_dispatcher` to dispatch to a functor like the other algorithms.  You can pass `p_src->type_` as the vertex type, INT32 as the edge type (we don't do anything with edges here), `p_weights->type_` as the weight type (default to `FLOAT32` if no weight) and default store transposed and multi gpu to `FALSE` (also not used).

That would let you handle any arbitrary combination of types in a simple call.

You should probably add a `CAPI_EXPECTS` check to verify that src and dst are the same type.

Reply:
Right. I had it hardcoded to certain type just to be able to move on. That's why I added a FIXME and I eventually came back to fix it with the `vertex_dispatcher`

> INT32 as the edge type (we don't do anything with edges here)

Actually this should not be hardcoded to INT32 because if we do so, we won't be able to support INT64 `src` and `dst`.